casper/src/main/scala/coop/rchain/casper/EstimatorHelper.scala

```diff
@@ -91,32 +127,132 @@ object EstimatorHelper {
+        .partition(isVolatile(_, allConsumeEvents, allProduceEvents))
+      producesInVolatileCommEvents = volatileCommEvents.flatMap(_.produces)
+      consumesInVolatileCommEvents = volatileCommEvents.map(_.consume)
+      produceEvents                = allProduceEvents.filterNot(producesInVolatileCommEvents.contains(_))
+      consumeEvents                = allConsumeEvents.filterNot(consumesInVolatileCommEvents.contains(_))
+    } yield (produceEvents, consumeEvents, nonVolatileCommEvents)
+
+  private[this] def extractJoinedChannels(b: BlockEvents): Set[Blake2b256Hash] = {
+    def joinedChannels(consumes: Set[Consume]) =
+      consumes.withFilter(Consume.hasJoins).flatMap(_.channelsHashes)
+    b match {
+      case (_, consumes, comms) =>
+        joinedChannels(consumes) ++ joinedChannels(comms.map(_.consume))
+    }
+  }
 
-      producesInCommEvents = commEvents.flatMap(_.produces)
-      consumeInCommEvents  = commEvents.map(_.consume)
+  private[this] def tuplespaceEventsPerChannel(
+      b: BlockEvents
+  ): Map[Blake2b256Hash, Set[TuplespaceEvent]] =
+    b match {
+      case (produces, consumes, comms) =>
+        val nonPersistentProducesInComms = comms.flatMap(_.produces).filterNot(_.persistent)
+        val nonPersistentConsumesInComms = comms.map(_.consume).filterNot(_.persistent)
+        val freeProduces                 = produces.filterNot(nonPersistentProducesInComms.contains)
+        val freeConsumes                 = consumes.filterNot(nonPersistentConsumesInComms.contains)
+        val produceEvents = freeProduces
+          .map(TuplespaceEvent.from(_))
 
-      ancestorProduceChannels = produceEvents
-        .filterNot(producesInCommEvents.contains(_))
-        .map(_.channelsHash)
-        .toSet
+        val consumeEvents = freeConsumes
+          .flatMap(TuplespaceEvent.from(_))
 
-      ancestorConsumeChannels = consumeEvents
-        .filterNot(consumeInCommEvents.contains(_))
-        .flatMap(_.channelsHashes)
-        .toSet
+        val commEvents = comms
+          .flatMap(TuplespaceEvent.from(_, produces))
 
-      ancestorCommChannels = commEvents
-        .filterNot(isVolatile(_, consumeEvents, produceEvents))
-        .flatMap { comm =>
-          comm.consume.channelsHashes ++ comm.produces.map(_.channelsHash)
+        (produceEvents
+          .combine(consumeEvents)
+          .combine(commEvents))
+          .groupBy(_._1)
+          .mapValues[Set[TuplespaceEvent]](_.map(_._2))
+    }
+
+  final case class TuplespaceEvent(
+      incoming: TuplespaceOperation,
+      matched: Option[TuplespaceOperation]
+  )
+  final case class TuplespaceOperation(
+      polarity: Polarity,
+      cardinality: Cardinality,
+      eventHash: Blake2b256Hash
+  )
+
+  trait Polarity
+  case object Send    extends Polarity
+  case object Receive extends Polarity
+
+  trait Cardinality
+  case object Linear    extends Cardinality
+  case object NonLinear extends Cardinality
+
+  object TuplespaceEvent {
```

Comment:
That's a lot of new code that we should extract to a separate file, and think of ways to separate it from EstimatorHelper for better testability maybe (less coupled setup -> faster tets)

Reply:
Good idea, I would rather do that in a follow up PR though if you don't mind